tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/setup-connector.md

```diff
@@ -0,0 +1,607 @@
+}
+{"level":30,"time":1705491544625,"pid":47901,"hostname":"Seans-MBP.lan","reqId":"req-6","res":{"statusCode":500},"responseTime":0.37700000405311584,"msg":"request completed"}
+```
+
+In the logs of the app, we can see the request that was sent. It identifies the name of the collection, and a query 
+object to run. The query has a list of fields to retrieve, and a limit of 10 rows. With this as a guide, we can 
+start to implement our query function.
+
+## Query
+
+The query function is going to delegate to a function called `fetch_rows`, but only when rows are requested, which is
+indicated by the presence of the query fields property.
+
+```typescript
+async function query(configuration: RawConfiguration, state: State, request: QueryRequest): Promise<QueryResponse> {
+  console.log(JSON.stringify(request, null, 2));
+  const rows = request.query.fields && await fetch_rows(state, request);
+  return [{ rows }];
+}
+```
+
+Later, we'll also implement aggregates here.
+
+Let's define the `fetch_rows` function:
+
+```typescript
+async function fetch_rows(state: State, request: QueryRequest): Promise<{
+  [k: string]: RowFieldValue
+}[]> {
+  const fields = [];
+
+  for (const fieldName in request.query.fields) {
+    if (Object.prototype.hasOwnProperty.call(request.query.fields, fieldName)) {
+      const field = request.query.fields[fieldName];
+      switch (field.type) {
+        case 'column':
+          fields.push(`${field.column} AS ${fieldName}`);
+          break;
+        case 'relationship':
+          throw new Error("Relationships are not supported");
+      }
+
+    }
+  }
+
+  if (request.query.order_by != null) {
+    throw new NotSupported("Sorting is not supported");
+  }
+
+  const limit_clause = request.query.limit == null ? "" : `LIMIT ${request.query.limit}`;
+  const offset_clause = request.query.offset == null ? "" : `OFFSET ${request.query.offset}`;
+
+  const sql = `SELECT ${fields.join(", ")} FROM ${request.collection} ${limit_clause} ${offset_clause}`;
+
+  console.log(JSON.stringify({ sql }, null, 2));
+
+  return state.db.all(sql);
+}
+```
+This function breaks down the request that we saw earlier and produces some SQL with this basic shape here. The
+requested fields get pushed down in the target list, and the limit and offset clauses are generated based on the
+request as well. Notice that we don't fetch more data than we need, either in terms of rows or columns. That's the
+benefit of connectors - we get to push down the query execution to the data sources themselves.
+
+## Test again
+
+Now let's see it work in the test runner. We'll rebuild and restart the connector, and run the tests again.
+
+Of course, we still see our tests fail, but now we've made some progress because the most basic tests are passing. If we
+look at the connector logs, we can see that we're now receiving some more advanced queries which we're not handling yet,
+such as queries with predicates and orderings.
```

Comment:
Again, as a user, I'm having trouble understanding where and when things should be failing. For the connector logs, it might make sense to pull out a single request and response that's now working that wasn't working before. That way, a user could easily grep through the logs and see what's chagned.

Reply:
Resolved